luarules/gadgets/unit_custom_weapons_overpen.lua

```diff
@@ -0,0 +1,449 @@
+        damageType = weaponData.damages[unitArmorType[unitDefID]]
+    elseif featureID then
+        health, healthMax = Spring.GetFeatureHealth(unitID)
+        damageType = weaponData.damages[0]
+    end
+    local damageBase = min(damage, damageType)
+
+    if weaponData.falloff then
+        damage = damage * damageLeft
+        damageBase = damageBase * damageLeft
+        damageLeft = damageLeft - health / damageBase - weaponData.penalty
+
+        if weaponData.expDefID and damageLeft <= explodeThreshold then
+            explodeDriver(projID, weaponData.expDefID, attackID, unitID, featureID)
+            return damage
+        end
+    end
+
+    if damageBase > health and damageBase >= healthMax * damageThreshold then
+        if damageLeft > 0 then
+            if weaponData.falloff then
+                driver[2] = damageLeft
+            end
+            respawn[projID] = driver
+        end
+        return damage
+    elseif weaponData.expDefID then
+        explodeDriver(projID, weaponData.expDefID, attackID, unitID, featureID)
+        return damage
+    end
+
+    local damageDone = min(1, damageBase / damageType)
+    return damage, inertia(damageDone) * damageDone
+end
+
+---Simulate the overpen effect by creating a new projectile.
+local function spawnPenetrator(projID, attackID, penDefID, unitID, featureID)
+    local penetrator = respawn[projID]
+    respawn[projID] = nil
+
+    local px, py, pz = spGetProjectilePosition(projID)
+    local timeToLive = spGetProjectileTimeToLive(projID)
+    local vx, vy, vz, vw = spGetProjectileVelocity(projID)
+
+    local driverData = penetrator[1]
+    local explodeID = driverData.expDefID
+
+    if driverData.slowing ~= nil then
+        local speedLeft = inertia(penetrator[2])
+        vx, vy, vz, vw = vx * speedLeft,
+                         vy * speedLeft,
+                         vz * speedLeft,
+                         vw * speedLeft
+    end
+
+    if driverData.penDefID ~= nil then
+        -- Spawn an alternative weapondef:
+        penDefID = driverData.penDefID
+        timeToLive = timeToLive * driverData.ttlRatio
+        local velRatio = driverData.velRatio
+        vx, vy, vz, vw = vx * velRatio,
+                         vy * velRatio,
+                         vz * velRatio,
+                         vw * velRatio
+        -- Penetrators may or may not be drivers, themselves:
+        penetrator[1] = weaponParams[penDefID] -- nil if not
+    end
+
+    local mx, my, mz, radius
+    if unitID ~= nil then
+        local _
+        _, _, _, mx, my, mz = spGetUnitPosition(unitID, true)
+        radius = spGetUnitRadius(unitID)
+    elseif featureID ~= nil then
+        local _
+        _, _, _, mx, my, mz = Spring.GetFeaturePosition(featureID, true)
+        radius = Spring.GetFeatureRadius(featureID)
+    end
+
+    -- Get the time to travel to a position opposite the target's sphere collider.
+    local frames = (radius / vw) * (cos(atan((mx - px) / radius - vx / vw)) +
```

Comment:
Can velocity ever be 0?

Reply:
A stationary projectile that a unit runs into sounds like an error in a scripted MoveCtrl. Still, that's a possible issue.